### PR TITLE
Admin abilities

### PR DESCRIPTION
**Pull request by ericpullukaran**

### CODE REVIEW

File: src/lib/auth/actions.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ const AUTH_REDIRECT_ORIGIN =
   env.NODE_ENV === "production" ? "https://promptracer.dev" : "http://localhost:3000"
 const AUTH_REDIRECT_URL = `${AUTH_REDIRECT_ORIGIN}/auth/callback`
 
-export const loginWithGitHubAction = action.action(async ({}) => {
+export const loginWithGitHubAction = action.action(async ({ }) => {
```
Comment (soorria) [inline src/lib/auth/actions.ts]: is this your format on save?

definitely gotta figure out what's going on

File: src/lib/auth/profile.ts [hunk 2]
```diff
@@ -35,6 +36,7 @@ export async function upsertProfile(authUser: User) {
     .values({
       id: authUser.id,
       name: getNameFromAuthUser(authUser),
+      role: !!authUser.email && ADMIN_EMAILS.includes(authUser.email) ? 'admin' : 'user',
```
Comment (soorria) [inline src/lib/auth/profile.ts]: imo let's remove this and make admin-ness manual (via drizzle)

File: src/lib/auth/user.ts [hunk 3]
```diff
@@ -3,17 +3,26 @@ import { cache } from "react"
...
   const sb = createServerClient()
   const { data, error } = await sb.auth.getUser()
 
-  if (error) {
+  if (data.user === null || error) {
+    logger.error(error, "failed to get user")
+    return null
+  }
+
+  const dbUser = await getDBUser(data.user.id);
```
Comment (soorria) [inline src/lib/auth/user.ts]: may not want to do this since now it makes every usage of `getAuthUser` require a db query

Comment (soorria) [reply src/lib/auth/user.ts]: if we need to check if a user is admin, we can make it a separate function

File: src/types/supabase-override.d.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Need to scope the import to enable overriding the types
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import * as supabase from '@supabase/supabase-js'
+
+declare module '@supabase/supabase-js' {
```
Comment (soorria) [inline src/types/supabase-override.d.ts]: pls remove

File: src/lib/games/trpc.ts [hunk 5]
```diff
@@ -268,14 +268,53 @@ export const gameRouter = createTRPCRouter({
...
+        throw new TRPCError({ code: "NOT_FOUND", message: "Game not found" })
+      }
+
+      if (!input.game_state) {
+        throw new TRPCError({ code: "BAD_REQUEST", message: "Must specify desired game state" })
+      }
+
+      if (input.game_state === "finished") {
+        await cancelInngestGameWorkflow(ctx.inngest, game.id)
+        finalizeGame(input.game_id)
```
Comment (soorria) [inline src/lib/games/trpc.ts]: need to await this

Comment (soorria) [reply src/lib/games/trpc.ts]: also may want to return early

File: src/lib/games/trpc.ts [hunk 6]
```diff
@@ -268,14 +268,53 @@ export const gameRouter = createTRPCRouter({
...
+
+      if (ctx.user.role !== 'admin') {
+        throw new TRPCError({ code: "FORBIDDEN", message: "Only admins can force progress games" })
+      }
+
+      if (!game) {
+        throw new TRPCError({ code: "NOT_FOUND", message: "Game not found" })
+      }
+
+      if (!input.game_state) {
```
Comment (soorria) [inline src/lib/games/trpc.ts]: can remove this & make game_state not optional in the input

File: src/lib/auth/profile.ts (hunk 2 above)
Comment (ericpullukaran) [reply src/lib/auth/profile.ts]: @soorria wdym here 👁️ 👁️

File: src/lib/auth/actions.ts (hunk 1 above)
Comment (ericpullukaran) [reply src/lib/auth/actions.ts]: yeaah lmao I wonder if my prettier settings are scuffed

File: src/lib/auth/user.ts (hunk 3 above)
Comment (ericpullukaran) [reply src/lib/auth/user.ts]: good shout

File: src/lib/auth/actions.ts (hunk 1 above)
Comment (ericpullukaran) [reply src/lib/auth/actions.ts]: fixed this turns out it was my default formatter got reset 💀

**Pull request merged**